cmd/query/app/apiv3/otlp_translator.go

```diff
@@ -0,0 +1,364 @@
+		tags = append(tags, model.String(conventions.AttributeServiceName, process.GetServiceName()))
+	}
+	return &resourcev1.Resource{
+		Attributes: jTagsToOTLP(tags, nil),
+	}
+}
+
+func jTagsToOTLP(tags []model.KeyValue, ignoreKeys map[string]bool) []*commonv1.KeyValue {
+	var kvs []*commonv1.KeyValue
+	for _, tag := range tags {
+		if ignoreKeys[tag.GetKey()] {
+			continue
+		}
+
+		kv := &commonv1.KeyValue{
+			Key:   tag.GetKey(),
+			Value: &commonv1.AnyValue{},
+		}
+		switch tag.GetVType() {
+		case model.ValueType_STRING:
+			kv.Value.Value = &commonv1.AnyValue_StringValue{
+				StringValue: tag.GetVStr(),
+			}
+		case model.ValueType_BOOL:
+			kv.Value.Value = &commonv1.AnyValue_BoolValue{
+				BoolValue: tag.GetVBool(),
+			}
+		case model.ValueType_INT64:
+			kv.Value.Value = &commonv1.AnyValue_IntValue{
+				IntValue: tag.GetVInt64(),
+			}
+		case model.ValueType_FLOAT64:
+			kv.Value.Value = &commonv1.AnyValue_DoubleValue{
+				DoubleValue: tag.GetVFloat64(),
+			}
+		case model.ValueType_BINARY:
+			kv.Value.Value = &commonv1.AnyValue_BytesValue{
+				BytesValue: tag.GetVBinary(),
+			}
+		default:
+			kv.Value.Value = &commonv1.AnyValue_StringValue{
+				StringValue: tag.String(),
+			}
+		}
+		kvs = append(kvs, kv)
+	}
+	return kvs
+}
+
+func jLogsToOTLP(logs []model.Log) []*v1.Span_Event {
+	events := make([]*v1.Span_Event, len(logs))
+	for i, l := range logs {
+
+		var name string
+		var ignoreKeys map[string]bool
+		if messageTag, ok := model.KeyValues(l.GetFields()).FindByKey(tracetranslator.TagMessage); ok {
+			name = messageTag.GetVStr()
+			ignoreKeys = map[string]bool{}
+			ignoreKeys[tracetranslator.TagMessage] = true
+		}
+
+		events[i] = &v1.Span_Event{
+			TimeUnixNano: uint64(l.GetTimestamp().UnixNano()),
+			Name:         name,
+			Attributes:   jTagsToOTLP(l.GetFields(), ignoreKeys),
+		}
+	}
+	return events
+}
+
+func jReferencesToOTLP(refs []model.SpanRef, excludeParentID model.SpanID) []*v1.Span_Link {
+	if len(refs) == 0 || len(refs) == 1 && refs[0].SpanID == excludeParentID && refs[0].RefType == model.ChildOf {
+		return nil
+	}
+	var links []*v1.Span_Link
+	for _, r := range refs {
+		if r.SpanID == excludeParentID && r.GetRefType() == model.ChildOf {
+			continue
+		}
+		links = append(links, &v1.Span_Link{
+			TraceId: uint64ToTraceID(r.TraceID.High, r.TraceID.Low),
+			SpanId:  uint64ToSpanID(uint64(r.SpanID)),
+		})
+	}
+
+	return links
+}
+
+func getSpanStatus(tags []model.KeyValue) (*v1.Status, map[string]bool) {
+	statusCode := v1.Status_STATUS_CODE_UNSET
+	statusMessage := ""
+	statusExists := false
+
+	ignoreKeys := map[string]bool{}
+	kvs := model.KeyValues(tags)
+	if _, ok := kvs.FindByKey(tracetranslator.TagError); ok {
+		statusCode = v1.Status_STATUS_CODE_ERROR
+		statusExists = true
+		ignoreKeys[tracetranslator.TagError] = true
+	}
+	if tag, ok := kvs.FindByKey(tracetranslator.TagStatusCode); ok {
+		statusExists = true
+		if code, err := getStatusCodeValFromTag(tag); err == nil {
```

Comment:
although I agree that we _should not_ return an error to the API caller or even return early in this function if there is an error in translating status codes, would it be worth logging (even on debug) the error so the problem can be addressed?

I imagine failure to map a jaeger span status code to OTLP status code should be more likely the exception than the norm, right?

Reply:
It should be an exception state, this is based on OTELcol translator which does not log the error either.